File: docs/js/charts/pestizide.js
/* ===========================================================================
   Biodiversitäts-Dashboard Österreich — Themenstrang: pestizide
   ---------------------------------------------------------------------------
   Wird nach js/kern.js geladen; die Helfer kommen aus window.BIO.
   =========================================================================== */
(function (BIO) {
"use strict";
const { stil, zahl, pz, basis, achse, tabelle, setzeText, setzeHtml,
        diagramme, schrift, hoverDunkler, istSchmal } = BIO;

/* --- Pestizidabsatz und seine Aufteilung ------------------------------
   Gestapelte Balken je Jahr, darüber die Gesamtlinie. Die Balken tragen
   die Aufteilung, die Linie den Verlauf — und sie schließt die Lücke
   2015, für die Eurostat nur den Gesamtwert veröffentlicht.

   WARUM DIE AUFTEILUNG ÜBERHAUPT IM BILD STEHT: Gegen die Aussage
   „der Absatz steigt" kommt sofort der Einwand, das seien vor allem
   anorganische Fungizide — Schwefel und Kupfer, die schwer wiegen und
   auch im Biolandbau zugelassen sind. Der Einwand gehört nicht in eine
   Fußnote, sondern ins Bild: Dort ist zu sehen, dass beide Teile wachsen.

   UND DASS ER ANDERS AUSGEHT ALS ERWARTET, ist der eigentliche Befund.
   Die anorganischen Fungizide wachsen SCHNELLER als der Rest (+81,5 %
   gegen +43,5 %), ihr Anteil steigt von 21,8 auf 26,0 %. Die Planungsakte
   hielt das Gegenteil fest — sie hatte „anorganisch" als Kupfer plus
   Schwefel gerechnet und dabei die dritte Unterkategorie übersehen, die
   sich verzwölffacht hat. Der Text sagt das ausdrücklich, statt die
   frühere Fassung stillschweigend zu ersetzen.

   KEINE ZIELMARKE BEI −50 %: Das Ziel der Farm-to-Fork-Strategie misst
   Einsatz und Risiko über einen gewichteten Indikator, nicht den Absatz
   in Kilogramm, und die Verordnung, die es verbindlich gemacht hätte, ist
   2023 gescheitert. Eine Marke wäre falscher Maßstab und kein geltendes
   Ziel zugleich. Dieselbe Linie wie bei `schutzgebiete`, wo die
   30-%-Marke seit 04.09.2026 „EU-weites Ziel" heißt.

   ZUR FARBWAHL: zwei Serientöne, keine Ampel. Anorganisch und übrig sind
   Stoffgruppen, keine Bewertungsstufen — `--viz-kritisch` bleibt draußen
   (Konvention: nur für Status). Die Linie nimmt `--viz-text-2`, damit sie
   sich von beiden Balkenteilen absetzt und nicht als dritte Kategorie
   gelesen wird. */

const FARBEN = {
  anorganisch: "--viz-series-2",
  rest:        "--viz-series-1",
};

function bauePestizide(daten) {
  const S = schrift();
  if (!daten?.punkte?.length) return;
  const abschnitt = document.getElementById("s-pestizide");
  if (abschnitt) abschnitt.style.display = "";

  const feld = document.getElementById("c-pestizide");
  if (!feld) return;
  const d = echarts.getInstanceByDom(feld) || echarts.init(feld, null, { renderer: "svg" });
  if (!diagramme.includes(d)) diagramme.push(d);

  const punkte = daten.punkte;
  const jahre = punkte.map((p) => String(p.jahr));

  setzeText("u-pestizide",
    `In Verkehr gebrachte Wirkstoffmenge in Tonnen · Österreich · ` +
    `${daten.beginn} bis ${daten.stand}`);
  setzeText("h-pestizide", daten.hinweis ?? "");

  d.setOption({
    ...basis(),
    grid: { left: 8, right: istSchmal(feld) ? 16 : 52, top: 40, bottom: 8,
            containLabel: true },
    legend: {
      top: 0, left: "center", icon: "roundRect",
      itemWidth: 11, itemHeight: 11, itemGap: 14,
      data: ["übrige Mittel", "anorganische Fungizide", "Absatz insgesamt"],
      textStyle: { color: stil("--viz-text-2"), fontSize: S.serie },
    },
    tooltip: {
      ...basis().tooltip, trigger: "axis",
      axisPointer: { type: "shadow", shadowStyle: { color: stil("--viz-grid"), opacity: 0.35 } },
      /* Die Kopfzeile nennt den Gesamtwert. Bei 2015 sagt der Tooltip
         ausdrücklich, dass die Aufteilung fehlt — sonst liest sich die
         Lücke als Null. */
      formatter: (p) => {
        const z = punkte[p[0].dataIndex];
        const kopf = `<strong>${z.jahr}</strong><br><strong>${zahl(z.gesamt)}</strong> Tonnen`;
        if (z.anorganisch === null) {
          return kopf + `<br><span style="color:${stil("--viz-muted")}">` +
            `Aufteilung für dieses Jahr vertraulich</span>`;
        }
        return kopf +
          `&nbsp;&nbsp;<span style="color:${stil("--viz-muted")}">` +
          `${pz(z.anteil, 1)} % anorganische Fungizide</span><br>` +
          `anorganische Fungizide&nbsp;&nbsp;<strong>${zahl(z.anorganisch)}</strong> t<br>` +
          `übrige Mittel&nbsp;&nbsp;<strong>${zahl(z.rest)}</strong> t`;
      },
    },
    xAxis: { ...achse(), type: "category", data: jahre, splitLine: { show: false },
      axisLabel: { hideOverlap: true, color: stil("--viz-muted"), fontSize: S.achse } },
    yAxis: { ...achse(), type: "value", axisLine: { show: false },
      axisLabel: { hideOverlap: true, color: stil("--viz-muted"),
                   fontSize: S.achse, formatter: (v) => zahl(v) } },
    series: [
      {
        name: "anorganische Fungizide", type: "bar", stack: "absatz",
        data: punkte.map((p) => p.anorganisch),
        itemStyle: { color: stil(FARBEN.anorganisch) },
        emphasis: hoverDunkler(stil(FARBEN.anorganisch)),
      },
      {
        name: "übrige Mittel", type: "bar", stack: "absatz",
        data: punkte.map((p) => p.rest),
        itemStyle: { color: stil(FARBEN.rest), borderRadius: [4, 4, 0, 0] },
        emphasis: hoverDunkler(stil(FARBEN.rest)),
      },
      /* Die Linie läuft über ALLE Jahre, auch über 2015, wo die Balken
         fehlen. Ohne sie wäre dort ein Loch und der Gesamtwert dieses
         Jahres — 3.778 t — gar nicht abzulesen. */
      {
        name: "Absatz insgesamt", type: "line", smooth: false, z: 3,
        /* KEIN `endLabel`. Am 04.09.2026 am ausgelieferten Stand gemessen:
           ECharts setzt ihn hier an den ERSTEN Punkt und zeigt dessen Wert —
           auf Pages bei `schutzgebiete` (27,5 statt 29,3), `vogel` und
           `falter` genauso. Vermutlich hängt er am Aufklapp-Clip der
           Einblendung, der nicht läuft, wenn der Abschnitt beim Bau noch
           `display:none` trägt.

           Stattdessen trägt der letzte Datenpunkt sein Etikett selbst. Das
           ist an das Datum gebunden und nicht an eine Animation — es kann
           gar nicht an der falschen Stelle landen. */
        data: punkte.map((p) => p.gesamt),
        markPoint: BIO.endEtikett(punkte.map((p) => p.gesamt), feld,
          (r) => zahl(r.value) + " t"),
        lineStyle: { color: stil("--viz-text-2"), width: 2 },
        itemStyle: { color: stil("--viz-text-2") },
        symbol: "circle", symbolSize: 5,
      },
    ],
  }, { replaceMerge: ["series", "xAxis", "yAxis", "legend"] });

  /* Die Notiz trägt vier Dinge, die die Balken nicht zeigen: dass die
     Kurve über dem Berg ist, wie der naheliegende Einwand ausgeht, warum
     ein Jahr keine Aufteilung hat, und warum hier keine Zielmarke steht. */
  const richtung = daten.schneller === "anorganisch"
    ? `Der Einwand geht anders aus als erwartet: Die anorganischen Fungizide — ` +
      `Schwefel, Kupfer und Verwandtes, auch im Biolandbau zugelassen — wachsen ` +
      `mit <strong>${pz(daten.wachstum_anorganisch, 1)} Prozent</strong> ` +
      `<em>schneller</em> als die übrigen Mittel (${pz(daten.wachstum_rest, 1)} Prozent). ` +
      `Ihr Anteil am Absatz steigt von ${pz(daten.anteil_beginn, 1)} auf ` +
      `${pz(daten.anteil_aktuell, 1)} Prozent.`
    : `Die anorganischen Fungizide wachsen mit ${pz(daten.wachstum_anorganisch, 1)} ` +
      `Prozent langsamer als die übrigen Mittel (${pz(daten.wachstum_rest, 1)} Prozent); ` +
      `der Anstieg liegt also im übrigen Teil.`;

  setzeHtml("n-pestizide",
    `Der Absatz ist nicht durchgehend gestiegen: Sein Höchststand liegt ` +
    `${daten.hoehepunkt_jahr} bei <strong>${zahl(daten.hoehepunkt)} Tonnen</strong>, ` +
    `seither sind es ${pz(Math.abs(daten.seit_hoehepunkt), 1)} Prozent weniger. ` +
    richtung + " " +
    (daten.ohne_teilung?.length
      ? `Für ${daten.ohne_teilung.join(", ")} veröffentlicht Eurostat nur den ` +
        `Gesamtwert, die Aufteilung ist dort vertraulich — deshalb fehlen die ` +
        `Balken und läuft nur die Linie durch. `
      : "") +
    `Das oft zitierte EU-Ziel „minus 50 Prozent" ist hier bewusst nicht ` +
    `eingezeichnet: Es misst Einsatz und Risiko über einen gewichteten ` +
    `Indikator, nicht die abgesetzte Menge.`);

  setzeHtml("t-pestizide", tabelle(
    [{ titel: "Jahr", wert: (z) => z.jahr },
     { titel: "Absatz gesamt", num: true, wert: (z) => zahl(z.gesamt) + " t" },
     { titel: "anorganische Fungizide", num: true,
       wert: (z) => z.anorganisch === null ? "vertraulich" : zahl(z.anorganisch) + " t" },
     { titel: "übrige Mittel", num: true,
       wert: (z) => z.rest === null ? "vertraulich" : zahl(z.rest) + " t" },
     { titel: "Anteil anorganisch", num: true,
       wert: (z) => z.anteil === null ? "–" : pz(z.anteil, 1) + " %" }],
    punkte
  ));
}

BIO.bauePestizide = bauePestizide;
})(window.BIO);

File: etl/pestizide.py
"""
Pestizidabsatz Österreich — Eurostat aei_fm_salpest09.

Was die Reihe misst: die in Verkehr gebrachte Menge Wirkstoff in Kilogramm,
nicht die ausgebrachte und nicht ihre Wirkung. Ein Kilogramm Schwefel und ein
Kilogramm eines synthetischen Wirkstoffs wiegen hier gleich viel.

DER ANSTIEG UND DER NAHELIEGENDE EINWAND. Der Absatz liegt 2024 rund die
Hälfte über 2011. Der erste Einwand dagegen lautet: Das seien vor allem
anorganische Fungizide — Schwefel und Kupfer, die schwer wiegen und auch im
Biolandbau zugelassen sind. Deshalb teilt dieses Modul den Absatz auf und
zeigt beide Teile im Bild, statt den Einwand in einer Fußnote abzuräumen.

**Die Planungsakte lag hier falsch, am 04.09.2026 korrigiert.** Sie hielt
fest, der Anteil der anorganischen Mittel *sinke* (24,8 % → 19,5 %) und der
Anstieg liege im synthetischen Rest. Das gilt nur, wenn man „anorganisch" als
Kupfer PLUS Schwefel liest. Nimmt man die Gruppe, die Eurostat selbst so
nennt — `F01`, anorganische Fungizide —, dreht sich der Befund:

    Gesamtabsatz   3.448 t (2011) → 5.232 t (2024)   +51,7 %
    davon F01        750 t        → 1.362 t          +81,5 %
    Rest           2.698 t        → 3.870 t          +43,5 %
    Anteil F01      21,8 %        → 26,0 %           steigt

Der Unterschied zwischen beiden Lesarten ist eine einzige Unterkategorie:
`F01_99`, sonstige anorganische Fungizide, **28 t (2012) → 341 t (2024)**,
also das Zwölffache. Kupfer und Schwefel allein verlieren tatsächlich Anteil;
die Gruppe als Ganzes gewinnt ihn.

KEINE ZIELMARKE. Das oft zitierte EU-Ziel „−50 % bis 2030" gehört zur
Farm-to-Fork-Strategie und misst **Einsatz und Risiko** über einen
gewichteten Indikator, nicht den Absatz in Kilogramm. Die Verordnung, die es
verbindlich gemacht hätte, ist 2023 im Parlament gescheitert und danach
zurückgezogen worden. Eine Marke bei −50 % neben dieser Kurve wäre deshalb
zweimal falsch: falscher Maßstab und kein geltendes Ziel.

LÜCKE 2015: Für die Aufteilung ist 2015 vertraulich. Der Gesamtwert ist da,
die beiden Teile fehlen. Das Bild lässt die Lücke stehen, statt sie zu
überbrücken.
"""

from __future__ import annotations

import config
from gemeinsam import jsonstat_reihe, lade_json, log, quelle_vermerken, warnen


def _reihe(code: str, name: str) -> dict[str, float]:
    url = f"{config.EUROSTAT_BASIS}/{config.PESTIZID_CODE}"
    params = dict(config.PESTIZID_PARAMS, pesticid=code)
    return jsonstat_reihe(lade_json(url, params), f"aei_fm_salpest09 ({name})")


def baue_pestizide() -> dict | None:
    log("\n[21/21] Pestizidabsatz — Eurostat aei_fm_salpest09")

    gesamt = _reihe(config.PESTIZID_GESAMT, "Insgesamt")
    if not gesamt:
        warnen("Pestizide: keine Gesamtreihe — Abschnitt bleibt ausgeblendet")
        return None

    anorg = _reihe(config.PESTIZID_ANORGANISCH, "anorganische Fungizide")

    jahre = sorted(gesamt)
    erstes, letztes = jahre[0], jahre[-1]

    # --- Gegenprobe 1: keine Lücke in der Gesamtreihe ----------------------
    # Der Abschnitt behauptet einen Verlauf. Fehlt darin ein Jahr, ist die
    # Linie eine Interpolation, die niemand angekündigt hat.
    luecken = [str(j) for j in range(int(erstes), int(letztes) + 1) if str(j) not in gesamt]
    if luecken:
        warnen(
            f"Pestizide: Gesamtreihe hat Lücken bei {', '.join(luecken)} — "
            f"der Verlauf wäre keine durchgehende Messung"
        )
        return None

    # --- Gegenprobe 2: der Teil ist nie größer als das Ganze ---------------
    zu_gross = [j for j in anorg if j in gesamt and anorg[j] > gesamt[j]]
    if zu_gross:
        warnen(
            f"Pestizide: anorganische Fungizide übersteigen den Gesamtabsatz "
            f"({', '.join(sorted(zu_gross))}) — die Codes passen nicht zueinander"
        )
        return None

    # --- Gegenprobe 3: die drei Unterkategorien ergeben die Gruppe ---------
    # Kupfer + Schwefel + Sonstige müssen F01 auf das Kilogramm treffen. Das
    # prüft, dass `F01` wirklich die Obergruppe ist und nicht eine vierte,
    # gleichnamige Kategorie — und es ist die Rechnung, auf der die
    # Korrektur an der Planungsakte beruht.
    teile = {
        name: _reihe(code, name)
        for name, code in config.PESTIZID_TEILE.items()
    }
    abweichung = []
    for j in sorted(anorg):
        if all(j in t for t in teile.values()):
            summe = sum(t[j] for t in teile.values())
            if abs(summe - anorg[j]) > 1:
                abweichung.append(f"{j}: {summe:.0f} statt {anorg[j]:.0f}")
    if abweichung:
        warnen(
            "Pestizide: Unterkategorien summieren nicht auf F01 — "
            + "; ".join(abweichung[:3])
        )
        return None

    punkte = [
        {
            "jahr": int(j),
            "gesamt": round(gesamt[j] / 1000),
            "anorganisch": round(anorg[j] / 1000) if j in anorg else None,
            "rest": round((gesamt[j] - anorg[j]) / 1000) if j in anorg else None,
            "anteil": round(anorg[j] / gesamt[j] * 100, 1) if j in anorg else None,
        }
        for j in jahre
    ]

    ohne_teilung = [p["jahr"] for p in punkte if p["anorganisch"] is None]

    def wachstum(a: float, b: float) -> float:
        return round((b / a - 1) * 100, 1)

    hoch_jahr = max(gesamt, key=lambda j: gesamt[j])
    seit_hoch = wachstum(gesamt[hoch_jahr], gesamt[letztes])

    basis_jahre = [j for j in config.PESTIZID_BASISPERIODE if j in gesamt]
    basis = sum(gesamt[j] for j in basis_jahre) / len(basis_jahre) if basis_jahre else None

    hat_teilung = anorg and erstes in anorg and letztes in anorg
    w_gesamt = wachstum(gesamt[erstes], gesamt[letztes])
    w_anorg = wachstum(anorg[erstes], anorg[letztes]) if hat_teilung else None
    w_rest = (wachstum(gesamt[erstes] - anorg[erstes], gesamt[letztes] - anorg[letztes])
              if hat_teilung else None)

    # --- Gegenprobe 4: trägt der Satz, den die Notiz schreiben wird? -------
    # Die Notiz sagt, welcher Teil schneller wächst. Diese Aussage wird hier
    # aus den Zahlen abgeleitet und nicht im Frontend behauptet — sonst
    # überlebt sie den nächsten Jahrgang, in dem sie vielleicht falsch ist.
    schneller = None
    if hat_teilung:
        schneller = "anorganisch" if w_anorg > w_rest else "rest"

    log(f"    {erstes}: {gesamt[erstes]/1000:,.0f} t  →  {letztes}: "
        f"{gesamt[letztes]/1000:,.0f} t  ({w_gesamt:+.1f} %)")
    if hat_teilung:
        log(f"    davon anorganische Fungizide {w_anorg:+.1f} %, Rest {w_rest:+.1f} % "
            f"— schneller wächst: {schneller}")
    log(f"    Höhepunkt {hoch_jahr} ({gesamt[hoch_jahr]/1000:,.0f} t), seither {seit_hoch:+.1f} %")
    if ohne_teilung:
        log(f"    ohne Aufteilung (vertraulich): {ohne_teilung}")

    quelle_vermerken(
        name="Eurostat — aei_fm_salpest09, Absatz von Pflanzenschutzmitteln",
        url="https://ec.europa.eu/eurostat/databrowser/view/aei_fm_salpest09",
        lizenz="Eurostat-Nutzungsbedingungen",
        stand=str(letztes),
        art="api",
    )

    return {
        "punkte": punkte,
        "beginn": int(erstes),
        "stand": int(letztes),
        "gesamt_aktuell": round(gesamt[letztes] / 1000),
        "gesamt_beginn": round(gesamt[erstes] / 1000),
        "wachstum": w_gesamt,
        "wachstum_anorganisch": w_anorg,
        "wachstum_rest": w_rest,
        "schneller": schneller,
        "anteil_beginn": round(anorg[erstes] / gesamt[erstes] * 100, 1) if hat_teilung else None,
        "anteil_aktuell": round(anorg[letztes] / gesamt[letztes] * 100, 1) if hat_teilung else None,
        "hoehepunkt_jahr": int(hoch_jahr),
        "hoehepunkt": round(gesamt[hoch_jahr] / 1000),
        "seit_hoehepunkt": seit_hoch,
        "basisperiode": [int(j) for j in basis_jahre],
        "basis_mittel": round(basis / 1000) if basis else None,
        "gegen_basis": wachstum(basis, gesamt[letztes]) if basis else None,
        "ohne_teilung": ohne_teilung,
        "kachel_wert": w_gesamt,
        "hinweis": (
            "Gemessen wird die in Verkehr gebrachte Menge Wirkstoff, nicht die "
            "ausgebrachte und nicht ihre Wirkung: Ein Kilogramm Schwefel wiegt hier "
            "so viel wie ein Kilogramm eines synthetischen Mittels."
        ),
    }
